src/pal/src/synchobj/mutex.cpp

```diff
@@ -613,3 +775,777 @@ DWORD SPINLOCKTryAcquire (LONG * lock)
+        *createdRef = created;
+    }
+    if (created)
+    {
+        // If the shared memory file was created, the creation/deletion file lock would have been acquired so that we can
+        // initialize the shared data
+        autoCleanup.m_acquiredCreationDeletionFileLock = true;
+    }
+    if (processDataHeader == nullptr)
+    {
+        _ASSERTE(!createIfNotExist);
+        return nullptr;
+    }
+    autoCleanup.m_processDataHeader = processDataHeader;
+
+    if (created)
+    {
+        // Initialize the shared data
+        new(processDataHeader->GetSharedDataHeader()->GetData()) NamedMutexSharedData;
+    }
+
+    if (processDataHeader->GetData() == nullptr)
+    {
+    #if !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+        // Create the lock files directory
+        char lockFilePath[SHARED_MEMORY_MAX_FILE_PATH_CHAR_COUNT + 1];
+        SIZE_T lockFilePathCharCount =
+            SharedMemoryHelpers::CopyString(lockFilePath, 0, SHARED_MEMORY_LOCK_FILES_DIRECTORY_PATH);
+        if (created)
+        {
+            SharedMemoryHelpers::EnsureDirectoryExists(lockFilePath);
+        }
+
+        // Create the session directory
+        lockFilePath[lockFilePathCharCount++] = '/';
+        SharedMemoryId *id = processDataHeader->GetId();
+        lockFilePathCharCount = id->AppendSessionDirectoryName(lockFilePath, lockFilePathCharCount);
+        if (created)
+        {
+            SharedMemoryHelpers::EnsureDirectoryExists(lockFilePath);
+            autoCleanup.m_lockFilePath = lockFilePath;
+            autoCleanup.m_sessionDirectoryPathCharCount = lockFilePathCharCount;
+        }
+
+        // Create or open the lock file
+        lockFilePath[lockFilePathCharCount++] = '/';
+        lockFilePathCharCount =
+            SharedMemoryHelpers::CopyString(lockFilePath, lockFilePathCharCount, id->GetName(), id->GetNameCharCount());
+        int lockFileDescriptor = SharedMemoryHelpers::CreateOrOpenFile(lockFilePath, created);
+        if (lockFileDescriptor == -1)
+        {
+            _ASSERTE(!created);
+            if (createIfNotExist)
+            {
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+            }
+            return nullptr;
+        }
+        autoCleanup.m_createdLockFile = created;
+        autoCleanup.m_lockFileDescriptor = lockFileDescriptor;
+    #endif // !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+
+        // Create the process data
+        void *processDataBuffer = SharedMemoryHelpers::Alloc(sizeof(NamedMutexProcessData));
+        NamedMutexProcessData *processData;
+        try
+        {
+            processData =
+                new(processDataBuffer)
+                NamedMutexProcessData(
+                    processDataHeader
+                #if !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+                    ,
+                    lockFileDescriptor
+                #endif // !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+                );
+        }
+        catch (...)
```

Comment:
When handing an exception, I would agree it's important to catch a specific type of exception. But when doing resource cleanup and rethrowing, where the cleanup code is not specific to any particular type of exception and needs to happen regardless of the type of exception, the code should reflect that.